// @google-cloud/bigtable/build/src/util/mock-servers/service-implementations/bigtable-client-mock-service.js
"use strict";
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.BigtableAdminClientMockService = exports.BigtableClientMockService = void 0;
const grpc = require("@grpc/grpc-js");
const jsonProtos = require("../../../../protos/protos.json");
const protoLoader = require("@grpc/proto-loader");
const mock_service_1 = require("../mock-service");
const packageDefinition = protoLoader.fromJSON(jsonProtos, {
    keepCase: true,
    longs: String,
    enums: String,
    defaults: true,
    oneofs: true,
});
const proto = grpc.loadPackageDefinition(packageDefinition);
class BigtableClientMockService extends mock_service_1.MockService {
    service = 
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    proto.google.bigtable.v2.Bigtable.service;
}
exports.BigtableClientMockService = BigtableClientMockService;
class BigtableAdminClientMockService extends mock_service_1.MockService {
    service = 
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    proto.google.bigtable.admin.v2.BigtableInstanceAdmin.service;
}
exports.BigtableAdminClientMockService = BigtableAdminClientMockService;
