// @google-cloud/bigtable/build/src/v2/bigtable_instance_admin_client.js
"use strict";
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.BigtableInstanceAdminClient = void 0;
const jsonProtos = require("../../protos/protos.json");
const google_gax_1 = require("google-gax");
/**
 * Client JSON configuration object, loaded from
 * `src/v2/bigtable_instance_admin_client_config.json`.
 * This file defines retry strategy and timeouts for all API methods in this library.
 */
const gapicConfig = require("./bigtable_instance_admin_client_config.json");
const version = require('../../../package.json').version;
/**
 *  Service for creating, configuring, and deleting Cloud Bigtable Instances and
 *  Clusters. Provides access to the Instance and Cluster schemas only, not the
 *  tables' metadata or data stored in those tables.
 * @class
 * @memberof v2
 */
class BigtableInstanceAdminClient {
    _terminated = false;
    _opts;
    _providedCustomServicePath;
    _gaxModule;
    _gaxGrpc;
    _protos;
    _defaults;
    _universeDomain;
    _servicePath;
    _log = google_gax_1.loggingUtils.log('bigtable');
    auth;
    descriptors = {
        page: {},
        stream: {},
        longrunning: {},
        batching: {},
    };
    warn;
    innerApiCalls;
    pathTemplates;
    operationsClient;
    bigtableInstanceAdminStub;
    /**
     * Construct an instance of BigtableInstanceAdminClient.
     *
     * @param {object} [options] - The configuration object.
     * The options accepted by the constructor are described in detail
     * in [this document](https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#creating-the-client-instance).
     * The common options are:
     * @param {object} [options.credentials] - Credentials object.
     * @param {string} [options.credentials.client_email]
     * @param {string} [options.credentials.private_key]
     * @param {string} [options.email] - Account email address. Required when
     *     using a .pem or .p12 keyFilename.
     * @param {string} [options.keyFilename] - Full path to the a .json, .pem, or
     *     .p12 key downloaded from the Google Developers Console. If you provide
     *     a path to a JSON file, the projectId option below is not necessary.
     *     NOTE: .pem and .p12 require you to specify options.email as well.
     * @param {number} [options.port] - The port on which to connect to
     *     the remote host.
     * @param {string} [options.projectId] - The project ID from the Google
     *     Developer's Console, e.g. 'grape-spaceship-123'. We will also check
     *     the environment variable GCLOUD_PROJECT for your project ID. If your
     *     app is running in an environment which supports
     *     {@link https://cloud.google.com/docs/authentication/application-default-credentials Application Default Credentials},
     *     your project ID will be detected automatically.
     * @param {string} [options.apiEndpoint] - The domain name of the
     *     API remote host.
     * @param {gax.ClientConfig} [options.clientConfig] - Client configuration override.
     *     Follows the structure of {@link gapicConfig}.
     * @param {boolean} [options.fallback] - Use HTTP/1.1 REST mode.
     *     For more information, please check the
     *     {@link https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#http11-rest-api-mode documentation}.
     * @param {gax} [gaxInstance]: loaded instance of `google-gax`. Useful if you
     *     need to avoid loading the default gRPC version and want to use the fallback
     *     HTTP implementation. Load only fallback version and pass it to the constructor:
     *     ```
     *     const gax = require('google-gax/build/src/fallback'); // avoids loading google-gax with gRPC
     *     const client = new BigtableInstanceAdminClient({fallback: true}, gax);
     *     ```
     */
    constructor(opts, gaxInstance) {
        // Ensure that options include all the required fields.
        const staticMembers = this
            .constructor;
        if (opts?.universe_domain &&
            opts?.universeDomain &&
            opts?.universe_domain !== opts?.universeDomain) {
            throw new Error('Please set either universe_domain or universeDomain, but not both.');
        }
        const universeDomainEnvVar = typeof process === 'object' && typeof process.env === 'object'
            ? process.env['GOOGLE_CLOUD_UNIVERSE_DOMAIN']
            : undefined;
        this._universeDomain =
            opts?.universeDomain ??
                opts?.universe_domain ??
                universeDomainEnvVar ??
                'googleapis.com';
        this._servicePath = 'bigtableadmin.' + this._universeDomain;
        const servicePath = opts?.servicePath || opts?.apiEndpoint || this._servicePath;
        this._providedCustomServicePath = !!(opts?.servicePath || opts?.apiEndpoint);
        const port = opts?.port || staticMembers.port;
        const clientConfig = opts?.clientConfig ?? {};
        const fallback = opts?.fallback ??
            (typeof window !== 'undefined' && typeof window?.fetch === 'function');
        opts = Object.assign({ servicePath, port, clientConfig, fallback }, opts);
        // Request numeric enum values if REST transport is used.
        opts.numericEnums = true;
        // If scopes are unset in options and we're connecting to a non-default endpoint, set scopes just in case.
        if (servicePath !== this._servicePath && !('scopes' in opts)) {
            opts['scopes'] = staticMembers.scopes;
        }
        // Load google-gax module synchronously if needed
        if (!gaxInstance) {
            gaxInstance = require('google-gax');
        }
        // Choose either gRPC or proto-over-HTTP implementation of google-gax.
        this._gaxModule = opts.fallback ? gaxInstance.fallback : gaxInstance;
        // Create a `gaxGrpc` object, with any grpc-specific options sent to the client.
        this._gaxGrpc = new this._gaxModule.GrpcClient(opts);
        // Save options to use in initialize() method.
        this._opts = opts;
        // Save the auth object to the client, for use by other methods.
        this.auth = this._gaxGrpc.auth;
        // Set useJWTAccessWithScope on the auth object.
        this.auth.useJWTAccessWithScope = true;
        // Set defaultServicePath on the auth object.
        this.auth.defaultServicePath = this._servicePath;
        // Set the default scopes in auth client if needed.
        if (servicePath === this._servicePath) {
            this.auth.defaultScopes = staticMembers.scopes;
        }
        // Determine the client header string.
        const clientHeader = [`gax/${this._gaxModule.version}`, `gapic/${version}`];
        if (typeof process === 'object' && 'versions' in process) {
            clientHeader.push(`gl-node/${process.versions.node}`);
        }
        else {
            clientHeader.push(`gl-web/${this._gaxModule.version}`);
        }
        if (!opts.fallback) {
            clientHeader.push(`grpc/${this._gaxGrpc.grpcVersion}`);
        }
        else {
            clientHeader.push(`rest/${this._gaxGrpc.grpcVersion}`);
        }
        if (opts.libName && opts.libVersion) {
            clientHeader.push(`${opts.libName}/${opts.libVersion}`);
        }
        // Load the applicable protos.
        this._protos = this._gaxGrpc.loadProtoJSON(jsonProtos);
        // This API contains "path templates"; forward-slash-separated
        // identifiers to uniquely identify resources within the API.
        // Create useful helper objects for these.
        this.pathTemplates = {
            appProfilePathTemplate: new this._gaxModule.PathTemplate('projects/{project}/instances/{instance}/appProfiles/{app_profile}'),
            authorizedViewPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/instances/{instance}/tables/{table}/authorizedViews/{authorized_view}'),
            backupPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/instances/{instance}/clusters/{cluster}/backups/{backup}'),
            clusterPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/instances/{instance}/clusters/{cluster}'),
            hotTabletPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/instances/{instance}/clusters/{cluster}/hotTablets/{hot_tablet}'),
            instancePathTemplate: new this._gaxModule.PathTemplate('projects/{project}/instances/{instance}'),
            logicalViewPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/instances/{instance}/logicalViews/{logical_view}'),
            materializedViewPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/instances/{instance}/materializedViews/{materialized_view}'),
            projectPathTemplate: new this._gaxModule.PathTemplate('projects/{project}'),
            snapshotPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/instances/{instance}/clusters/{cluster}/snapshots/{snapshot}'),
            tablePathTemplate: new this._gaxModule.PathTemplate('projects/{project}/instances/{instance}/tables/{table}'),
        };
        // Some of the methods on this service return "paged" results,
        // (e.g. 50 results at a time, with tokens to get subsequent
        // pages). Denote the keys used for pagination and results.
        this.descriptors.page = {
            listAppProfiles: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'appProfiles'),
            listHotTablets: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'hotTablets'),
            listLogicalViews: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'logicalViews'),
            listMaterializedViews: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'materializedViews'),
        };
        const protoFilesRoot = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
        // This API contains "long-running operations", which return a
        // an Operation object that allows for tracking of the operation,
        // rather than holding a request open.
        const lroOptions = {
            auth: this.auth,
            grpc: 'grpc' in this._gaxGrpc ? this._gaxGrpc.grpc : undefined,
        };
        if (opts.fallback) {
            lroOptions.protoJson = protoFilesRoot;
            lroOptions.httpRules = [
                {
                    selector: 'google.longrunning.Operations.CancelOperation',
                    post: '/v2/{name=operations/**}:cancel',
                },
                {
                    selector: 'google.longrunning.Operations.DeleteOperation',
                    delete: '/v2/{name=operations/**}',
                },
                {
                    selector: 'google.longrunning.Operations.GetOperation',
                    get: '/v2/{name=operations/**}',
                },
                {
                    selector: 'google.longrunning.Operations.ListOperations',
                    get: '/v2/{name=operations/projects/**}/operations',
                },
            ];
        }
        this.operationsClient = this._gaxModule
            .lro(lroOptions)
            .operationsClient(opts);
        const createInstanceResponse = protoFilesRoot.lookup('.google.bigtable.admin.v2.Instance');
        const createInstanceMetadata = protoFilesRoot.lookup('.google.bigtable.admin.v2.CreateInstanceMetadata');
        const partialUpdateInstanceResponse = protoFilesRoot.lookup('.google.bigtable.admin.v2.Instance');
        const partialUpdateInstanceMetadata = protoFilesRoot.lookup('.google.bigtable.admin.v2.UpdateInstanceMetadata');
        const createClusterResponse = protoFilesRoot.lookup('.google.bigtable.admin.v2.Cluster');
        const createClusterMetadata = protoFilesRoot.lookup('.google.bigtable.admin.v2.CreateClusterMetadata');
        const updateClusterResponse = protoFilesRoot.lookup('.google.bigtable.admin.v2.Cluster');
        const updateClusterMetadata = protoFilesRoot.lookup('.google.bigtable.admin.v2.UpdateClusterMetadata');
        const partialUpdateClusterResponse = protoFilesRoot.lookup('.google.bigtable.admin.v2.Cluster');
        const partialUpdateClusterMetadata = protoFilesRoot.lookup('.google.bigtable.admin.v2.PartialUpdateClusterMetadata');
        const updateAppProfileResponse = protoFilesRoot.lookup('.google.bigtable.admin.v2.AppProfile');
        const updateAppProfileMetadata = protoFilesRoot.lookup('.google.bigtable.admin.v2.UpdateAppProfileMetadata');
        const createLogicalViewResponse = protoFilesRoot.lookup('.google.bigtable.admin.v2.LogicalView');
        const createLogicalViewMetadata = protoFilesRoot.lookup('.google.bigtable.admin.v2.CreateLogicalViewMetadata');
        const updateLogicalViewResponse = protoFilesRoot.lookup('.google.bigtable.admin.v2.LogicalView');
        const updateLogicalViewMetadata = protoFilesRoot.lookup('.google.bigtable.admin.v2.UpdateLogicalViewMetadata');
        const createMaterializedViewResponse = protoFilesRoot.lookup('.google.bigtable.admin.v2.MaterializedView');
        const createMaterializedViewMetadata = protoFilesRoot.lookup('.google.bigtable.admin.v2.CreateMaterializedViewMetadata');
        const updateMaterializedViewResponse = protoFilesRoot.lookup('.google.bigtable.admin.v2.MaterializedView');
        const updateMaterializedViewMetadata = protoFilesRoot.lookup('.google.bigtable.admin.v2.UpdateMaterializedViewMetadata');
        this.descriptors.longrunning = {
            createInstance: new this._gaxModule.LongrunningDescriptor(this.operationsClient, createInstanceResponse.decode.bind(createInstanceResponse), createInstanceMetadata.decode.bind(createInstanceMetadata)),
            partialUpdateInstance: new this._gaxModule.LongrunningDescriptor(this.operationsClient, partialUpdateInstanceResponse.decode.bind(partialUpdateInstanceResponse), partialUpdateInstanceMetadata.decode.bind(partialUpdateInstanceMetadata)),
            createCluster: new this._gaxModule.LongrunningDescriptor(this.operationsClient, createClusterResponse.decode.bind(createClusterResponse), createClusterMetadata.decode.bind(createClusterMetadata)),
            updateCluster: new this._gaxModule.LongrunningDescriptor(this.operationsClient, updateClusterResponse.decode.bind(updateClusterResponse), updateClusterMetadata.decode.bind(updateClusterMetadata)),
            partialUpdateCluster: new this._gaxModule.LongrunningDescriptor(this.operationsClient, partialUpdateClusterResponse.decode.bind(partialUpdateClusterResponse), partialUpdateClusterMetadata.decode.bind(partialUpdateClusterMetadata)),
            updateAppProfile: new this._gaxModule.LongrunningDescriptor(this.operationsClient, updateAppProfileResponse.decode.bind(updateAppProfileResponse), updateAppProfileMetadata.decode.bind(updateAppProfileMetadata)),
            createLogicalView: new this._gaxModule.LongrunningDescriptor(this.operationsClient, createLogicalViewResponse.decode.bind(createLogicalViewResponse), createLogicalViewMetadata.decode.bind(createLogicalViewMetadata)),
            updateLogicalView: new this._gaxModule.LongrunningDescriptor(this.operationsClient, updateLogicalViewResponse.decode.bind(updateLogicalViewResponse), updateLogicalViewMetadata.decode.bind(updateLogicalViewMetadata)),
            createMaterializedView: new this._gaxModule.LongrunningDescriptor(this.operationsClient, createMaterializedViewResponse.decode.bind(createMaterializedViewResponse), createMaterializedViewMetadata.decode.bind(createMaterializedViewMetadata)),
            updateMaterializedView: new this._gaxModule.LongrunningDescriptor(this.operationsClient, updateMaterializedViewResponse.decode.bind(updateMaterializedViewResponse), updateMaterializedViewMetadata.decode.bind(updateMaterializedViewMetadata)),
        };
        // Put together the default options sent with requests.
        this._defaults = this._gaxGrpc.constructSettings('google.bigtable.admin.v2.BigtableInstanceAdmin', gapicConfig, opts.clientConfig || {}, { 'x-goog-api-client': clientHeader.join(' ') });
        // Set up a dictionary of "inner API calls"; the core implementation
        // of calling the API is handled in `google-gax`, with this code
        // merely providing the destination and request information.
        this.innerApiCalls = {};
        // Add a warn function to the client constructor so it can be easily tested.
        this.warn = this._gaxModule.warn;
    }
    /**
     * Initialize the client.
     * Performs asynchronous operations (such as authentication) and prepares the client.
     * This function will be called automatically when any class method is called for the
     * first time, but if you need to initialize it before calling an actual method,
     * feel free to call initialize() directly.
     *
     * You can await on this method if you want to make sure the client is initialized.
     *
     * @returns {Promise} A promise that resolves to an authenticated service stub.
     */
    initialize() {
        // If the client stub promise is already initialized, return immediately.
        if (this.bigtableInstanceAdminStub) {
            return this.bigtableInstanceAdminStub;
        }
        // Put together the "service stub" for
        // google.bigtable.admin.v2.BigtableInstanceAdmin.
        this.bigtableInstanceAdminStub = this._gaxGrpc.createStub(this._opts.fallback
            ? this._protos.lookupService('google.bigtable.admin.v2.BigtableInstanceAdmin')
            : // eslint-disable-next-line @typescript-eslint/no-explicit-any
                this._protos.google.bigtable.admin.v2.BigtableInstanceAdmin, this._opts, this._providedCustomServicePath);
        // Iterate over each of the methods that the service provides
        // and create an API call method for each.
        const bigtableInstanceAdminStubMethods = [
            'createInstance',
            'getInstance',
            'listInstances',
            'updateInstance',
            'partialUpdateInstance',
            'deleteInstance',
            'createCluster',
            'getCluster',
            'listClusters',
            'updateCluster',
            'partialUpdateCluster',
            'deleteCluster',
            'createAppProfile',
            'getAppProfile',
            'listAppProfiles',
            'updateAppProfile',
            'deleteAppProfile',
            'getIamPolicy',
            'setIamPolicy',
            'testIamPermissions',
            'listHotTablets',
            'createLogicalView',
            'getLogicalView',
            'listLogicalViews',
            'updateLogicalView',
            'deleteLogicalView',
            'createMaterializedView',
            'getMaterializedView',
            'listMaterializedViews',
            'updateMaterializedView',
            'deleteMaterializedView',
        ];
        for (const methodName of bigtableInstanceAdminStubMethods) {
            const callPromise = this.bigtableInstanceAdminStub.then(stub => (...args) => {
                if (this._terminated) {
                    return Promise.reject('The client has already been closed.');
                }
                const func = stub[methodName];
                return func.apply(stub, args);
            }, (err) => () => {
                throw err;
            });
            const descriptor = this.descriptors.page[methodName] ||
                this.descriptors.longrunning[methodName] ||
                undefined;
            const apiCall = this._gaxModule.createApiCall(callPromise, this._defaults[methodName], descriptor, this._opts.fallback);
            this.innerApiCalls[methodName] = apiCall;
        }
        return this.bigtableInstanceAdminStub;
    }
    /**
     * The DNS address for this API service.
     * @deprecated Use the apiEndpoint method of the client instance.
     * @returns {string} The DNS address for this service.
     */
    static get servicePath() {
        if (typeof process === 'object' &&
            typeof process.emitWarning === 'function') {
            process.emitWarning('Static servicePath is deprecated, please use the instance method instead.', 'DeprecationWarning');
        }
        return 'bigtableadmin.googleapis.com';
    }
    /**
     * The DNS address for this API service - same as servicePath.
     * @deprecated Use the apiEndpoint method of the client instance.
     * @returns {string} The DNS address for this service.
     */
    static get apiEndpoint() {
        if (typeof process === 'object' &&
            typeof process.emitWarning === 'function') {
            process.emitWarning('Static apiEndpoint is deprecated, please use the instance method instead.', 'DeprecationWarning');
        }
        return 'bigtableadmin.googleapis.com';
    }
    /**
     * The DNS address for this API service.
     * @returns {string} The DNS address for this service.
     */
    get apiEndpoint() {
        return this._servicePath;
    }
    get universeDomain() {
        return this._universeDomain;
    }
    /**
     * The port for this API service.
     * @returns {number} The default port for this service.
     */
    static get port() {
        return 443;
    }
    /**
     * The scopes needed to make gRPC calls for every method defined
     * in this service.
     * @returns {string[]} List of default scopes.
     */
    static get scopes() {
        return [
            'https://www.googleapis.com/auth/bigtable.admin',
            'https://www.googleapis.com/auth/bigtable.admin.cluster',
            'https://www.googleapis.com/auth/bigtable.admin.instance',
            'https://www.googleapis.com/auth/cloud-bigtable.admin',
            'https://www.googleapis.com/auth/cloud-bigtable.admin.cluster',
            'https://www.googleapis.com/auth/cloud-platform',
            'https://www.googleapis.com/auth/cloud-platform.read-only',
        ];
    }
    /**
     * Return the project ID used by this class.
     * @returns {Promise} A promise that resolves to string containing the project ID.
     */
    getProjectId(callback) {
        if (callback) {
            this.auth.getProjectId(callback);
            return;
        }
        return this.auth.getProjectId();
    }
    getInstance(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: request.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('getInstance request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getInstance response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .getInstance(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getInstance response %j', response);
            return [response, options, rawResponse];
        });
    }
    listInstances(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('listInstances request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('listInstances response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .listInstances(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('listInstances response %j', response);
            return [response, options, rawResponse];
        });
    }
    updateInstance(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: request.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('updateInstance request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('updateInstance response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .updateInstance(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('updateInstance response %j', response);
            return [response, options, rawResponse];
        });
    }
    deleteInstance(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: request.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('deleteInstance request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('deleteInstance response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .deleteInstance(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('deleteInstance response %j', response);
            return [response, options, rawResponse];
        });
    }
    getCluster(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: request.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('getCluster request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getCluster response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .getCluster(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getCluster response %j', response);
            return [response, options, rawResponse];
        });
    }
    listClusters(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('listClusters request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('listClusters response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .listClusters(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('listClusters response %j', response);
            return [response, options, rawResponse];
        });
    }
    deleteCluster(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: request.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('deleteCluster request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('deleteCluster response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .deleteCluster(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('deleteCluster response %j', response);
            return [response, options, rawResponse];
        });
    }
    createAppProfile(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('createAppProfile request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('createAppProfile response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .createAppProfile(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('createAppProfile response %j', response);
            return [response, options, rawResponse];
        });
    }
    getAppProfile(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: request.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('getAppProfile request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getAppProfile response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .getAppProfile(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getAppProfile response %j', response);
            return [response, options, rawResponse];
        });
    }
    deleteAppProfile(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: request.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('deleteAppProfile request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('deleteAppProfile response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .deleteAppProfile(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('deleteAppProfile response %j', response);
            return [response, options, rawResponse];
        });
    }
    getIamPolicy(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                resource: request.resource ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('getIamPolicy request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getIamPolicy response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .getIamPolicy(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getIamPolicy response %j', response);
            return [response, options, rawResponse];
        });
    }
    setIamPolicy(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                resource: request.resource ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('setIamPolicy request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('setIamPolicy response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .setIamPolicy(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('setIamPolicy response %j', response);
            return [response, options, rawResponse];
        });
    }
    testIamPermissions(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                resource: request.resource ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('testIamPermissions request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('testIamPermissions response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .testIamPermissions(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('testIamPermissions response %j', response);
            return [response, options, rawResponse];
        });
    }
    getLogicalView(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: request.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('getLogicalView request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getLogicalView response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .getLogicalView(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getLogicalView response %j', response);
            return [response, options, rawResponse];
        });
    }
    deleteLogicalView(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: request.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('deleteLogicalView request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('deleteLogicalView response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .deleteLogicalView(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('deleteLogicalView response %j', response);
            return [response, options, rawResponse];
        });
    }
    getMaterializedView(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: request.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('getMaterializedView request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getMaterializedView response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .getMaterializedView(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getMaterializedView response %j', response);
            return [response, options, rawResponse];
        });
    }
    deleteMaterializedView(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: request.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('deleteMaterializedView request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('deleteMaterializedView response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .deleteMaterializedView(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('deleteMaterializedView response %j', response);
            return [response, options, rawResponse];
        });
    }
    createInstance(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        const wrappedCallback = callback
            ? (error, response, rawResponse, _) => {
                this._log.info('createInstance response %j', rawResponse);
                callback(error, response, rawResponse, _); // We verified callback above.
            }
            : undefined;
        this._log.info('createInstance request %j', request);
        return this.innerApiCalls
            .createInstance(request, options, wrappedCallback)
            ?.then(([response, rawResponse, _]) => {
            this._log.info('createInstance response %j', rawResponse);
            return [response, rawResponse, _];
        });
    }
    /**
     * Check the status of the long running operation returned by `createInstance()`.
     * @param {String} name
     *   The operation name that will be passed.
     * @returns {Promise} - The promise which resolves to an object.
     *   The decoded operation object has result and metadata field to get information from.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#long-running-operations | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v2/bigtable_instance_admin.create_instance.js</caption>
     * region_tag:bigtableadmin_v2_generated_BigtableInstanceAdmin_CreateInstance_async
     */
    async checkCreateInstanceProgress(name) {
        this._log.info('createInstance long-running');
        const request = new this._gaxModule.operationsProtos.google.longrunning.GetOperationRequest({ name });
        const [operation] = await this.operationsClient.getOperation(request);
        const decodeOperation = new this._gaxModule.Operation(operation, this.descriptors.longrunning.createInstance, this._gaxModule.createDefaultBackoffSettings());
        return decodeOperation;
    }
    partialUpdateInstance(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'instance.name': request.instance.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        const wrappedCallback = callback
            ? (error, response, rawResponse, _) => {
                this._log.info('partialUpdateInstance response %j', rawResponse);
                callback(error, response, rawResponse, _); // We verified callback above.
            }
            : undefined;
        this._log.info('partialUpdateInstance request %j', request);
        return this.innerApiCalls
            .partialUpdateInstance(request, options, wrappedCallback)
            ?.then(([response, rawResponse, _]) => {
            this._log.info('partialUpdateInstance response %j', rawResponse);
            return [response, rawResponse, _];
        });
    }
    /**
     * Check the status of the long running operation returned by `partialUpdateInstance()`.
     * @param {String} name
     *   The operation name that will be passed.
     * @returns {Promise} - The promise which resolves to an object.
     *   The decoded operation object has result and metadata field to get information from.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#long-running-operations | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v2/bigtable_instance_admin.partial_update_instance.js</caption>
     * region_tag:bigtableadmin_v2_generated_BigtableInstanceAdmin_PartialUpdateInstance_async
     */
    async checkPartialUpdateInstanceProgress(name) {
        this._log.info('partialUpdateInstance long-running');
        const request = new this._gaxModule.operationsProtos.google.longrunning.GetOperationRequest({ name });
        const [operation] = await this.operationsClient.getOperation(request);
        const decodeOperation = new this._gaxModule.Operation(operation, this.descriptors.longrunning.partialUpdateInstance, this._gaxModule.createDefaultBackoffSettings());
        return decodeOperation;
    }
    createCluster(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        const wrappedCallback = callback
            ? (error, response, rawResponse, _) => {
                this._log.info('createCluster response %j', rawResponse);
                callback(error, response, rawResponse, _); // We verified callback above.
            }
            : undefined;
        this._log.info('createCluster request %j', request);
        return this.innerApiCalls
            .createCluster(request, options, wrappedCallback)
            ?.then(([response, rawResponse, _]) => {
            this._log.info('createCluster response %j', rawResponse);
            return [response, rawResponse, _];
        });
    }
    /**
     * Check the status of the long running operation returned by `createCluster()`.
     * @param {String} name
     *   The operation name that will be passed.
     * @returns {Promise} - The promise which resolves to an object.
     *   The decoded operation object has result and metadata field to get information from.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#long-running-operations | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v2/bigtable_instance_admin.create_cluster.js</caption>
     * region_tag:bigtableadmin_v2_generated_BigtableInstanceAdmin_CreateCluster_async
     */
    async checkCreateClusterProgress(name) {
        this._log.info('createCluster long-running');
        const request = new this._gaxModule.operationsProtos.google.longrunning.GetOperationRequest({ name });
        const [operation] = await this.operationsClient.getOperation(request);
        const decodeOperation = new this._gaxModule.Operation(operation, this.descriptors.longrunning.createCluster, this._gaxModule.createDefaultBackoffSettings());
        return decodeOperation;
    }
    updateCluster(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: request.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        const wrappedCallback = callback
            ? (error, response, rawResponse, _) => {
                this._log.info('updateCluster response %j', rawResponse);
                callback(error, response, rawResponse, _); // We verified callback above.
            }
            : undefined;
        this._log.info('updateCluster request %j', request);
        return this.innerApiCalls
            .updateCluster(request, options, wrappedCallback)
            ?.then(([response, rawResponse, _]) => {
            this._log.info('updateCluster response %j', rawResponse);
            return [response, rawResponse, _];
        });
    }
    /**
     * Check the status of the long running operation returned by `updateCluster()`.
     * @param {String} name
     *   The operation name that will be passed.
     * @returns {Promise} - The promise which resolves to an object.
     *   The decoded operation object has result and metadata field to get information from.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#long-running-operations | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v2/bigtable_instance_admin.update_cluster.js</caption>
     * region_tag:bigtableadmin_v2_generated_BigtableInstanceAdmin_UpdateCluster_async
     */
    async checkUpdateClusterProgress(name) {
        this._log.info('updateCluster long-running');
        const request = new this._gaxModule.operationsProtos.google.longrunning.GetOperationRequest({ name });
        const [operation] = await this.operationsClient.getOperation(request);
        const decodeOperation = new this._gaxModule.Operation(operation, this.descriptors.longrunning.updateCluster, this._gaxModule.createDefaultBackoffSettings());
        return decodeOperation;
    }
    partialUpdateCluster(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'cluster.name': request.cluster.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        const wrappedCallback = callback
            ? (error, response, rawResponse, _) => {
                this._log.info('partialUpdateCluster response %j', rawResponse);
                callback(error, response, rawResponse, _); // We verified callback above.
            }
            : undefined;
        this._log.info('partialUpdateCluster request %j', request);
        return this.innerApiCalls
            .partialUpdateCluster(request, options, wrappedCallback)
            ?.then(([response, rawResponse, _]) => {
            this._log.info('partialUpdateCluster response %j', rawResponse);
            return [response, rawResponse, _];
        });
    }
    /**
     * Check the status of the long running operation returned by `partialUpdateCluster()`.
     * @param {String} name
     *   The operation name that will be passed.
     * @returns {Promise} - The promise which resolves to an object.
     *   The decoded operation object has result and metadata field to get information from.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#long-running-operations | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v2/bigtable_instance_admin.partial_update_cluster.js</caption>
     * region_tag:bigtableadmin_v2_generated_BigtableInstanceAdmin_PartialUpdateCluster_async
     */
    async checkPartialUpdateClusterProgress(name) {
        this._log.info('partialUpdateCluster long-running');
        const request = new this._gaxModule.operationsProtos.google.longrunning.GetOperationRequest({ name });
        const [operation] = await this.operationsClient.getOperation(request);
        const decodeOperation = new this._gaxModule.Operation(operation, this.descriptors.longrunning.partialUpdateCluster, this._gaxModule.createDefaultBackoffSettings());
        return decodeOperation;
    }
    updateAppProfile(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'app_profile.name': request.appProfile.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        const wrappedCallback = callback
            ? (error, response, rawResponse, _) => {
                this._log.info('updateAppProfile response %j', rawResponse);
                callback(error, response, rawResponse, _); // We verified callback above.
            }
            : undefined;
        this._log.info('updateAppProfile request %j', request);
        return this.innerApiCalls
            .updateAppProfile(request, options, wrappedCallback)
            ?.then(([response, rawResponse, _]) => {
            this._log.info('updateAppProfile response %j', rawResponse);
            return [response, rawResponse, _];
        });
    }
    /**
     * Check the status of the long running operation returned by `updateAppProfile()`.
     * @param {String} name
     *   The operation name that will be passed.
     * @returns {Promise} - The promise which resolves to an object.
     *   The decoded operation object has result and metadata field to get information from.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#long-running-operations | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v2/bigtable_instance_admin.update_app_profile.js</caption>
     * region_tag:bigtableadmin_v2_generated_BigtableInstanceAdmin_UpdateAppProfile_async
     */
    async checkUpdateAppProfileProgress(name) {
        this._log.info('updateAppProfile long-running');
        const request = new this._gaxModule.operationsProtos.google.longrunning.GetOperationRequest({ name });
        const [operation] = await this.operationsClient.getOperation(request);
        const decodeOperation = new this._gaxModule.Operation(operation, this.descriptors.longrunning.updateAppProfile, this._gaxModule.createDefaultBackoffSettings());
        return decodeOperation;
    }
    createLogicalView(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        const wrappedCallback = callback
            ? (error, response, rawResponse, _) => {
                this._log.info('createLogicalView response %j', rawResponse);
                callback(error, response, rawResponse, _); // We verified callback above.
            }
            : undefined;
        this._log.info('createLogicalView request %j', request);
        return this.innerApiCalls
            .createLogicalView(request, options, wrappedCallback)
            ?.then(([response, rawResponse, _]) => {
            this._log.info('createLogicalView response %j', rawResponse);
            return [response, rawResponse, _];
        });
    }
    /**
     * Check the status of the long running operation returned by `createLogicalView()`.
     * @param {String} name
     *   The operation name that will be passed.
     * @returns {Promise} - The promise which resolves to an object.
     *   The decoded operation object has result and metadata field to get information from.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#long-running-operations | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v2/bigtable_instance_admin.create_logical_view.js</caption>
     * region_tag:bigtableadmin_v2_generated_BigtableInstanceAdmin_CreateLogicalView_async
     */
    async checkCreateLogicalViewProgress(name) {
        this._log.info('createLogicalView long-running');
        const request = new this._gaxModule.operationsProtos.google.longrunning.GetOperationRequest({ name });
        const [operation] = await this.operationsClient.getOperation(request);
        const decodeOperation = new this._gaxModule.Operation(operation, this.descriptors.longrunning.createLogicalView, this._gaxModule.createDefaultBackoffSettings());
        return decodeOperation;
    }
    updateLogicalView(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'logical_view.name': request.logicalView.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        const wrappedCallback = callback
            ? (error, response, rawResponse, _) => {
                this._log.info('updateLogicalView response %j', rawResponse);
                callback(error, response, rawResponse, _); // We verified callback above.
            }
            : undefined;
        this._log.info('updateLogicalView request %j', request);
        return this.innerApiCalls
            .updateLogicalView(request, options, wrappedCallback)
            ?.then(([response, rawResponse, _]) => {
            this._log.info('updateLogicalView response %j', rawResponse);
            return [response, rawResponse, _];
        });
    }
    /**
     * Check the status of the long running operation returned by `updateLogicalView()`.
     * @param {String} name
     *   The operation name that will be passed.
     * @returns {Promise} - The promise which resolves to an object.
     *   The decoded operation object has result and metadata field to get information from.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#long-running-operations | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v2/bigtable_instance_admin.update_logical_view.js</caption>
     * region_tag:bigtableadmin_v2_generated_BigtableInstanceAdmin_UpdateLogicalView_async
     */
    async checkUpdateLogicalViewProgress(name) {
        this._log.info('updateLogicalView long-running');
        const request = new this._gaxModule.operationsProtos.google.longrunning.GetOperationRequest({ name });
        const [operation] = await this.operationsClient.getOperation(request);
        const decodeOperation = new this._gaxModule.Operation(operation, this.descriptors.longrunning.updateLogicalView, this._gaxModule.createDefaultBackoffSettings());
        return decodeOperation;
    }
    createMaterializedView(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        const wrappedCallback = callback
            ? (error, response, rawResponse, _) => {
                this._log.info('createMaterializedView response %j', rawResponse);
                callback(error, response, rawResponse, _); // We verified callback above.
            }
            : undefined;
        this._log.info('createMaterializedView request %j', request);
        return this.innerApiCalls
            .createMaterializedView(request, options, wrappedCallback)
            ?.then(([response, rawResponse, _]) => {
            this._log.info('createMaterializedView response %j', rawResponse);
            return [response, rawResponse, _];
        });
    }
    /**
     * Check the status of the long running operation returned by `createMaterializedView()`.
     * @param {String} name
     *   The operation name that will be passed.
     * @returns {Promise} - The promise which resolves to an object.
     *   The decoded operation object has result and metadata field to get information from.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#long-running-operations | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v2/bigtable_instance_admin.create_materialized_view.js</caption>
     * region_tag:bigtableadmin_v2_generated_BigtableInstanceAdmin_CreateMaterializedView_async
     */
    async checkCreateMaterializedViewProgress(name) {
        this._log.info('createMaterializedView long-running');
        const request = new this._gaxModule.operationsProtos.google.longrunning.GetOperationRequest({ name });
        const [operation] = await this.operationsClient.getOperation(request);
        const decodeOperation = new this._gaxModule.Operation(operation, this.descriptors.longrunning.createMaterializedView, this._gaxModule.createDefaultBackoffSettings());
        return decodeOperation;
    }
    updateMaterializedView(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'materialized_view.name': request.materializedView.name ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        const wrappedCallback = callback
            ? (error, response, rawResponse, _) => {
                this._log.info('updateMaterializedView response %j', rawResponse);
                callback(error, response, rawResponse, _); // We verified callback above.
            }
            : undefined;
        this._log.info('updateMaterializedView request %j', request);
        return this.innerApiCalls
            .updateMaterializedView(request, options, wrappedCallback)
            ?.then(([response, rawResponse, _]) => {
            this._log.info('updateMaterializedView response %j', rawResponse);
            return [response, rawResponse, _];
        });
    }
    /**
     * Check the status of the long running operation returned by `updateMaterializedView()`.
     * @param {String} name
     *   The operation name that will be passed.
     * @returns {Promise} - The promise which resolves to an object.
     *   The decoded operation object has result and metadata field to get information from.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#long-running-operations | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v2/bigtable_instance_admin.update_materialized_view.js</caption>
     * region_tag:bigtableadmin_v2_generated_BigtableInstanceAdmin_UpdateMaterializedView_async
     */
    async checkUpdateMaterializedViewProgress(name) {
        this._log.info('updateMaterializedView long-running');
        const request = new this._gaxModule.operationsProtos.google.longrunning.GetOperationRequest({ name });
        const [operation] = await this.operationsClient.getOperation(request);
        const decodeOperation = new this._gaxModule.Operation(operation, this.descriptors.longrunning.updateMaterializedView, this._gaxModule.createDefaultBackoffSettings());
        return decodeOperation;
    }
    listAppProfiles(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listAppProfiles values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listAppProfiles request %j', request);
        return this.innerApiCalls
            .listAppProfiles(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listAppProfiles values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listAppProfiles`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The unique name of the instance for which a list of app profiles
     *   is requested. Values are of the form
     *   `projects/{project}/instances/{instance}`.
     *   Use `{instance} = '-'` to list AppProfiles for all Instances in a project,
     *   e.g., `projects/myproject/instances/-`.
     * @param {number} request.pageSize
     *   Maximum number of results per page.
     *
     *   A page_size of zero lets the server choose the number of items to return.
     *   A page_size which is strictly positive will return at most that many items.
     *   A negative page_size will cause an error.
     *
     *   Following the first request, subsequent paginated calls are not required
     *   to pass a page_size. If a page_size is set in subsequent calls, it must
     *   match the page_size given in the first request.
     * @param {string} request.pageToken
     *   The value of `next_page_token` returned by a previous call.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.bigtable.admin.v2.AppProfile|AppProfile} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listAppProfilesAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listAppProfilesStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        const defaultCallSettings = this._defaults['listAppProfiles'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('listAppProfiles stream %j', request);
        return this.descriptors.page.listAppProfiles.createStream(this.innerApiCalls.listAppProfiles, request, callSettings);
    }
    /**
     * Equivalent to `listAppProfiles`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The unique name of the instance for which a list of app profiles
     *   is requested. Values are of the form
     *   `projects/{project}/instances/{instance}`.
     *   Use `{instance} = '-'` to list AppProfiles for all Instances in a project,
     *   e.g., `projects/myproject/instances/-`.
     * @param {number} request.pageSize
     *   Maximum number of results per page.
     *
     *   A page_size of zero lets the server choose the number of items to return.
     *   A page_size which is strictly positive will return at most that many items.
     *   A negative page_size will cause an error.
     *
     *   Following the first request, subsequent paginated calls are not required
     *   to pass a page_size. If a page_size is set in subsequent calls, it must
     *   match the page_size given in the first request.
     * @param {string} request.pageToken
     *   The value of `next_page_token` returned by a previous call.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.bigtable.admin.v2.AppProfile|AppProfile}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v2/bigtable_instance_admin.list_app_profiles.js</caption>
     * region_tag:bigtableadmin_v2_generated_BigtableInstanceAdmin_ListAppProfiles_async
     */
    listAppProfilesAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        const defaultCallSettings = this._defaults['listAppProfiles'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('listAppProfiles iterate %j', request);
        return this.descriptors.page.listAppProfiles.asyncIterate(this.innerApiCalls['listAppProfiles'], request, callSettings);
    }
    listHotTablets(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listHotTablets values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listHotTablets request %j', request);
        return this.innerApiCalls
            .listHotTablets(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listHotTablets values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listHotTablets`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The cluster name to list hot tablets.
     *   Value is in the following form:
     *   `projects/{project}/instances/{instance}/clusters/{cluster}`.
     * @param {google.protobuf.Timestamp} request.startTime
     *   The start time to list hot tablets. The hot tablets in the response will
     *   have start times between the requested start time and end time. Start time
     *   defaults to Now if it is unset, and end time defaults to Now - 24 hours if
     *   it is unset. The start time should be less than the end time, and the
     *   maximum allowed time range between start time and end time is 48 hours.
     *   Start time and end time should have values between Now and Now - 14 days.
     * @param {google.protobuf.Timestamp} request.endTime
     *   The end time to list hot tablets.
     * @param {number} request.pageSize
     *   Maximum number of results per page.
     *
     *   A page_size that is empty or zero lets the server choose the number of
     *   items to return. A page_size which is strictly positive will return at most
     *   that many items. A negative page_size will cause an error.
     *
     *   Following the first request, subsequent paginated calls do not need a
     *   page_size field. If a page_size is set in subsequent calls, it must match
     *   the page_size given in the first request.
     * @param {string} request.pageToken
     *   The value of `next_page_token` returned by a previous call.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.bigtable.admin.v2.HotTablet|HotTablet} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listHotTabletsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listHotTabletsStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        const defaultCallSettings = this._defaults['listHotTablets'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('listHotTablets stream %j', request);
        return this.descriptors.page.listHotTablets.createStream(this.innerApiCalls.listHotTablets, request, callSettings);
    }
    /**
     * Equivalent to `listHotTablets`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The cluster name to list hot tablets.
     *   Value is in the following form:
     *   `projects/{project}/instances/{instance}/clusters/{cluster}`.
     * @param {google.protobuf.Timestamp} request.startTime
     *   The start time to list hot tablets. The hot tablets in the response will
     *   have start times between the requested start time and end time. Start time
     *   defaults to Now if it is unset, and end time defaults to Now - 24 hours if
     *   it is unset. The start time should be less than the end time, and the
     *   maximum allowed time range between start time and end time is 48 hours.
     *   Start time and end time should have values between Now and Now - 14 days.
     * @param {google.protobuf.Timestamp} request.endTime
     *   The end time to list hot tablets.
     * @param {number} request.pageSize
     *   Maximum number of results per page.
     *
     *   A page_size that is empty or zero lets the server choose the number of
     *   items to return. A page_size which is strictly positive will return at most
     *   that many items. A negative page_size will cause an error.
     *
     *   Following the first request, subsequent paginated calls do not need a
     *   page_size field. If a page_size is set in subsequent calls, it must match
     *   the page_size given in the first request.
     * @param {string} request.pageToken
     *   The value of `next_page_token` returned by a previous call.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.bigtable.admin.v2.HotTablet|HotTablet}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v2/bigtable_instance_admin.list_hot_tablets.js</caption>
     * region_tag:bigtableadmin_v2_generated_BigtableInstanceAdmin_ListHotTablets_async
     */
    listHotTabletsAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        const defaultCallSettings = this._defaults['listHotTablets'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('listHotTablets iterate %j', request);
        return this.descriptors.page.listHotTablets.asyncIterate(this.innerApiCalls['listHotTablets'], request, callSettings);
    }
    listLogicalViews(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listLogicalViews values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listLogicalViews request %j', request);
        return this.innerApiCalls
            .listLogicalViews(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listLogicalViews values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listLogicalViews`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The unique name of the instance for which the list of logical
     *   views is requested. Values are of the form
     *   `projects/{project}/instances/{instance}`.
     * @param {number} [request.pageSize]
     *   Optional. The maximum number of logical views to return. The service may
     *   return fewer than this value
     * @param {string} [request.pageToken]
     *   Optional. A page token, received from a previous `ListLogicalViews` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListLogicalViews` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.bigtable.admin.v2.LogicalView|LogicalView} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listLogicalViewsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listLogicalViewsStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        const defaultCallSettings = this._defaults['listLogicalViews'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('listLogicalViews stream %j', request);
        return this.descriptors.page.listLogicalViews.createStream(this.innerApiCalls.listLogicalViews, request, callSettings);
    }
    /**
     * Equivalent to `listLogicalViews`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The unique name of the instance for which the list of logical
     *   views is requested. Values are of the form
     *   `projects/{project}/instances/{instance}`.
     * @param {number} [request.pageSize]
     *   Optional. The maximum number of logical views to return. The service may
     *   return fewer than this value
     * @param {string} [request.pageToken]
     *   Optional. A page token, received from a previous `ListLogicalViews` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListLogicalViews` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.bigtable.admin.v2.LogicalView|LogicalView}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v2/bigtable_instance_admin.list_logical_views.js</caption>
     * region_tag:bigtableadmin_v2_generated_BigtableInstanceAdmin_ListLogicalViews_async
     */
    listLogicalViewsAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        const defaultCallSettings = this._defaults['listLogicalViews'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('listLogicalViews iterate %j', request);
        return this.descriptors.page.listLogicalViews.asyncIterate(this.innerApiCalls['listLogicalViews'], request, callSettings);
    }
    listMaterializedViews(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listMaterializedViews values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listMaterializedViews request %j', request);
        return this.innerApiCalls
            .listMaterializedViews(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listMaterializedViews values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listMaterializedViews`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The unique name of the instance for which the list of
     *   materialized views is requested. Values are of the form
     *   `projects/{project}/instances/{instance}`.
     * @param {number} [request.pageSize]
     *   Optional. The maximum number of materialized views to return. The service
     *   may return fewer than this value
     * @param {string} [request.pageToken]
     *   Optional. A page token, received from a previous `ListMaterializedViews`
     *   call. Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListMaterializedViews`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.bigtable.admin.v2.MaterializedView|MaterializedView} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listMaterializedViewsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listMaterializedViewsStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        const defaultCallSettings = this._defaults['listMaterializedViews'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('listMaterializedViews stream %j', request);
        return this.descriptors.page.listMaterializedViews.createStream(this.innerApiCalls.listMaterializedViews, request, callSettings);
    }
    /**
     * Equivalent to `listMaterializedViews`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The unique name of the instance for which the list of
     *   materialized views is requested. Values are of the form
     *   `projects/{project}/instances/{instance}`.
     * @param {number} [request.pageSize]
     *   Optional. The maximum number of materialized views to return. The service
     *   may return fewer than this value
     * @param {string} [request.pageToken]
     *   Optional. A page token, received from a previous `ListMaterializedViews`
     *   call. Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListMaterializedViews`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.bigtable.admin.v2.MaterializedView|MaterializedView}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v2/bigtable_instance_admin.list_materialized_views.js</caption>
     * region_tag:bigtableadmin_v2_generated_BigtableInstanceAdmin_ListMaterializedViews_async
     */
    listMaterializedViewsAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: request.parent ?? '',
            });
        const defaultCallSettings = this._defaults['listMaterializedViews'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('listMaterializedViews iterate %j', request);
        return this.descriptors.page.listMaterializedViews.asyncIterate(this.innerApiCalls['listMaterializedViews'], request, callSettings);
    }
    // --------------------
    // -- Path templates --
    // --------------------
    /**
     * Return a fully-qualified appProfile resource name string.
     *
     * @param {string} project
     * @param {string} instance
     * @param {string} app_profile
     * @returns {string} Resource name string.
     */
    appProfilePath(project, instance, appProfile) {
        return this.pathTemplates.appProfilePathTemplate.render({
            project: project,
            instance: instance,
            app_profile: appProfile,
        });
    }
    /**
     * Parse the project from AppProfile resource.
     *
     * @param {string} appProfileName
     *   A fully-qualified path representing AppProfile resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromAppProfileName(appProfileName) {
        return this.pathTemplates.appProfilePathTemplate.match(appProfileName)
            .project;
    }
    /**
     * Parse the instance from AppProfile resource.
     *
     * @param {string} appProfileName
     *   A fully-qualified path representing AppProfile resource.
     * @returns {string} A string representing the instance.
     */
    matchInstanceFromAppProfileName(appProfileName) {
        return this.pathTemplates.appProfilePathTemplate.match(appProfileName)
            .instance;
    }
    /**
     * Parse the app_profile from AppProfile resource.
     *
     * @param {string} appProfileName
     *   A fully-qualified path representing AppProfile resource.
     * @returns {string} A string representing the app_profile.
     */
    matchAppProfileFromAppProfileName(appProfileName) {
        return this.pathTemplates.appProfilePathTemplate.match(appProfileName)
            .app_profile;
    }
    /**
     * Return a fully-qualified authorizedView resource name string.
     *
     * @param {string} project
     * @param {string} instance
     * @param {string} table
     * @param {string} authorized_view
     * @returns {string} Resource name string.
     */
    authorizedViewPath(project, instance, table, authorizedView) {
        return this.pathTemplates.authorizedViewPathTemplate.render({
            project: project,
            instance: instance,
            table: table,
            authorized_view: authorizedView,
        });
    }
    /**
     * Parse the project from AuthorizedView resource.
     *
     * @param {string} authorizedViewName
     *   A fully-qualified path representing AuthorizedView resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromAuthorizedViewName(authorizedViewName) {
        return this.pathTemplates.authorizedViewPathTemplate.match(authorizedViewName).project;
    }
    /**
     * Parse the instance from AuthorizedView resource.
     *
     * @param {string} authorizedViewName
     *   A fully-qualified path representing AuthorizedView resource.
     * @returns {string} A string representing the instance.
     */
    matchInstanceFromAuthorizedViewName(authorizedViewName) {
        return this.pathTemplates.authorizedViewPathTemplate.match(authorizedViewName).instance;
    }
    /**
     * Parse the table from AuthorizedView resource.
     *
     * @param {string} authorizedViewName
     *   A fully-qualified path representing AuthorizedView resource.
     * @returns {string} A string representing the table.
     */
    matchTableFromAuthorizedViewName(authorizedViewName) {
        return this.pathTemplates.authorizedViewPathTemplate.match(authorizedViewName).table;
    }
    /**
     * Parse the authorized_view from AuthorizedView resource.
     *
     * @param {string} authorizedViewName
     *   A fully-qualified path representing AuthorizedView resource.
     * @returns {string} A string representing the authorized_view.
     */
    matchAuthorizedViewFromAuthorizedViewName(authorizedViewName) {
        return this.pathTemplates.authorizedViewPathTemplate.match(authorizedViewName).authorized_view;
    }
    /**
     * Return a fully-qualified backup resource name string.
     *
     * @param {string} project
     * @param {string} instance
     * @param {string} cluster
     * @param {string} backup
     * @returns {string} Resource name string.
     */
    backupPath(project, instance, cluster, backup) {
        return this.pathTemplates.backupPathTemplate.render({
            project: project,
            instance: instance,
            cluster: cluster,
            backup: backup,
        });
    }
    /**
     * Parse the project from Backup resource.
     *
     * @param {string} backupName
     *   A fully-qualified path representing Backup resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromBackupName(backupName) {
        return this.pathTemplates.backupPathTemplate.match(backupName).project;
    }
    /**
     * Parse the instance from Backup resource.
     *
     * @param {string} backupName
     *   A fully-qualified path representing Backup resource.
     * @returns {string} A string representing the instance.
     */
    matchInstanceFromBackupName(backupName) {
        return this.pathTemplates.backupPathTemplate.match(backupName).instance;
    }
    /**
     * Parse the cluster from Backup resource.
     *
     * @param {string} backupName
     *   A fully-qualified path representing Backup resource.
     * @returns {string} A string representing the cluster.
     */
    matchClusterFromBackupName(backupName) {
        return this.pathTemplates.backupPathTemplate.match(backupName).cluster;
    }
    /**
     * Parse the backup from Backup resource.
     *
     * @param {string} backupName
     *   A fully-qualified path representing Backup resource.
     * @returns {string} A string representing the backup.
     */
    matchBackupFromBackupName(backupName) {
        return this.pathTemplates.backupPathTemplate.match(backupName).backup;
    }
    /**
     * Return a fully-qualified cluster resource name string.
     *
     * @param {string} project
     * @param {string} instance
     * @param {string} cluster
     * @returns {string} Resource name string.
     */
    clusterPath(project, instance, cluster) {
        return this.pathTemplates.clusterPathTemplate.render({
            project: project,
            instance: instance,
            cluster: cluster,
        });
    }
    /**
     * Parse the project from Cluster resource.
     *
     * @param {string} clusterName
     *   A fully-qualified path representing Cluster resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromClusterName(clusterName) {
        return this.pathTemplates.clusterPathTemplate.match(clusterName).project;
    }
    /**
     * Parse the instance from Cluster resource.
     *
     * @param {string} clusterName
     *   A fully-qualified path representing Cluster resource.
     * @returns {string} A string representing the instance.
     */
    matchInstanceFromClusterName(clusterName) {
        return this.pathTemplates.clusterPathTemplate.match(clusterName).instance;
    }
    /**
     * Parse the cluster from Cluster resource.
     *
     * @param {string} clusterName
     *   A fully-qualified path representing Cluster resource.
     * @returns {string} A string representing the cluster.
     */
    matchClusterFromClusterName(clusterName) {
        return this.pathTemplates.clusterPathTemplate.match(clusterName).cluster;
    }
    /**
     * Return a fully-qualified hotTablet resource name string.
     *
     * @param {string} project
     * @param {string} instance
     * @param {string} cluster
     * @param {string} hot_tablet
     * @returns {string} Resource name string.
     */
    hotTabletPath(project, instance, cluster, hotTablet) {
        return this.pathTemplates.hotTabletPathTemplate.render({
            project: project,
            instance: instance,
            cluster: cluster,
            hot_tablet: hotTablet,
        });
    }
    /**
     * Parse the project from HotTablet resource.
     *
     * @param {string} hotTabletName
     *   A fully-qualified path representing HotTablet resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromHotTabletName(hotTabletName) {
        return this.pathTemplates.hotTabletPathTemplate.match(hotTabletName)
            .project;
    }
    /**
     * Parse the instance from HotTablet resource.
     *
     * @param {string} hotTabletName
     *   A fully-qualified path representing HotTablet resource.
     * @returns {string} A string representing the instance.
     */
    matchInstanceFromHotTabletName(hotTabletName) {
        return this.pathTemplates.hotTabletPathTemplate.match(hotTabletName)
            .instance;
    }
    /**
     * Parse the cluster from HotTablet resource.
     *
     * @param {string} hotTabletName
     *   A fully-qualified path representing HotTablet resource.
     * @returns {string} A string representing the cluster.
     */
    matchClusterFromHotTabletName(hotTabletName) {
        return this.pathTemplates.hotTabletPathTemplate.match(hotTabletName)
            .cluster;
    }
    /**
     * Parse the hot_tablet from HotTablet resource.
     *
     * @param {string} hotTabletName
     *   A fully-qualified path representing HotTablet resource.
     * @returns {string} A string representing the hot_tablet.
     */
    matchHotTabletFromHotTabletName(hotTabletName) {
        return this.pathTemplates.hotTabletPathTemplate.match(hotTabletName)
            .hot_tablet;
    }
    /**
     * Return a fully-qualified instance resource name string.
     *
     * @param {string} project
     * @param {string} instance
     * @returns {string} Resource name string.
     */
    instancePath(project, instance) {
        return this.pathTemplates.instancePathTemplate.render({
            project: project,
            instance: instance,
        });
    }
    /**
     * Parse the project from Instance resource.
     *
     * @param {string} instanceName
     *   A fully-qualified path representing Instance resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromInstanceName(instanceName) {
        return this.pathTemplates.instancePathTemplate.match(instanceName).project;
    }
    /**
     * Parse the instance from Instance resource.
     *
     * @param {string} instanceName
     *   A fully-qualified path representing Instance resource.
     * @returns {string} A string representing the instance.
     */
    matchInstanceFromInstanceName(instanceName) {
        return this.pathTemplates.instancePathTemplate.match(instanceName).instance;
    }
    /**
     * Return a fully-qualified logicalView resource name string.
     *
     * @param {string} project
     * @param {string} instance
     * @param {string} logical_view
     * @returns {string} Resource name string.
     */
    logicalViewPath(project, instance, logicalView) {
        return this.pathTemplates.logicalViewPathTemplate.render({
            project: project,
            instance: instance,
            logical_view: logicalView,
        });
    }
    /**
     * Parse the project from LogicalView resource.
     *
     * @param {string} logicalViewName
     *   A fully-qualified path representing LogicalView resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromLogicalViewName(logicalViewName) {
        return this.pathTemplates.logicalViewPathTemplate.match(logicalViewName)
            .project;
    }
    /**
     * Parse the instance from LogicalView resource.
     *
     * @param {string} logicalViewName
     *   A fully-qualified path representing LogicalView resource.
     * @returns {string} A string representing the instance.
     */
    matchInstanceFromLogicalViewName(logicalViewName) {
        return this.pathTemplates.logicalViewPathTemplate.match(logicalViewName)
            .instance;
    }
    /**
     * Parse the logical_view from LogicalView resource.
     *
     * @param {string} logicalViewName
     *   A fully-qualified path representing LogicalView resource.
     * @returns {string} A string representing the logical_view.
     */
    matchLogicalViewFromLogicalViewName(logicalViewName) {
        return this.pathTemplates.logicalViewPathTemplate.match(logicalViewName)
            .logical_view;
    }
    /**
     * Return a fully-qualified materializedView resource name string.
     *
     * @param {string} project
     * @param {string} instance
     * @param {string} materialized_view
     * @returns {string} Resource name string.
     */
    materializedViewPath(project, instance, materializedView) {
        return this.pathTemplates.materializedViewPathTemplate.render({
            project: project,
            instance: instance,
            materialized_view: materializedView,
        });
    }
    /**
     * Parse the project from MaterializedView resource.
     *
     * @param {string} materializedViewName
     *   A fully-qualified path representing MaterializedView resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromMaterializedViewName(materializedViewName) {
        return this.pathTemplates.materializedViewPathTemplate.match(materializedViewName).project;
    }
    /**
     * Parse the instance from MaterializedView resource.
     *
     * @param {string} materializedViewName
     *   A fully-qualified path representing MaterializedView resource.
     * @returns {string} A string representing the instance.
     */
    matchInstanceFromMaterializedViewName(materializedViewName) {
        return this.pathTemplates.materializedViewPathTemplate.match(materializedViewName).instance;
    }
    /**
     * Parse the materialized_view from MaterializedView resource.
     *
     * @param {string} materializedViewName
     *   A fully-qualified path representing MaterializedView resource.
     * @returns {string} A string representing the materialized_view.
     */
    matchMaterializedViewFromMaterializedViewName(materializedViewName) {
        return this.pathTemplates.materializedViewPathTemplate.match(materializedViewName).materialized_view;
    }
    /**
     * Return a fully-qualified project resource name string.
     *
     * @param {string} project
     * @returns {string} Resource name string.
     */
    projectPath(project) {
        return this.pathTemplates.projectPathTemplate.render({
            project: project,
        });
    }
    /**
     * Parse the project from Project resource.
     *
     * @param {string} projectName
     *   A fully-qualified path representing Project resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromProjectName(projectName) {
        return this.pathTemplates.projectPathTemplate.match(projectName).project;
    }
    /**
     * Return a fully-qualified snapshot resource name string.
     *
     * @param {string} project
     * @param {string} instance
     * @param {string} cluster
     * @param {string} snapshot
     * @returns {string} Resource name string.
     */
    snapshotPath(project, instance, cluster, snapshot) {
        return this.pathTemplates.snapshotPathTemplate.render({
            project: project,
            instance: instance,
            cluster: cluster,
            snapshot: snapshot,
        });
    }
    /**
     * Parse the project from Snapshot resource.
     *
     * @param {string} snapshotName
     *   A fully-qualified path representing Snapshot resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromSnapshotName(snapshotName) {
        return this.pathTemplates.snapshotPathTemplate.match(snapshotName).project;
    }
    /**
     * Parse the instance from Snapshot resource.
     *
     * @param {string} snapshotName
     *   A fully-qualified path representing Snapshot resource.
     * @returns {string} A string representing the instance.
     */
    matchInstanceFromSnapshotName(snapshotName) {
        return this.pathTemplates.snapshotPathTemplate.match(snapshotName).instance;
    }
    /**
     * Parse the cluster from Snapshot resource.
     *
     * @param {string} snapshotName
     *   A fully-qualified path representing Snapshot resource.
     * @returns {string} A string representing the cluster.
     */
    matchClusterFromSnapshotName(snapshotName) {
        return this.pathTemplates.snapshotPathTemplate.match(snapshotName).cluster;
    }
    /**
     * Parse the snapshot from Snapshot resource.
     *
     * @param {string} snapshotName
     *   A fully-qualified path representing Snapshot resource.
     * @returns {string} A string representing the snapshot.
     */
    matchSnapshotFromSnapshotName(snapshotName) {
        return this.pathTemplates.snapshotPathTemplate.match(snapshotName).snapshot;
    }
    /**
     * Return a fully-qualified table resource name string.
     *
     * @param {string} project
     * @param {string} instance
     * @param {string} table
     * @returns {string} Resource name string.
     */
    tablePath(project, instance, table) {
        return this.pathTemplates.tablePathTemplate.render({
            project: project,
            instance: instance,
            table: table,
        });
    }
    /**
     * Parse the project from Table resource.
     *
     * @param {string} tableName
     *   A fully-qualified path representing Table resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromTableName(tableName) {
        return this.pathTemplates.tablePathTemplate.match(tableName).project;
    }
    /**
     * Parse the instance from Table resource.
     *
     * @param {string} tableName
     *   A fully-qualified path representing Table resource.
     * @returns {string} A string representing the instance.
     */
    matchInstanceFromTableName(tableName) {
        return this.pathTemplates.tablePathTemplate.match(tableName).instance;
    }
    /**
     * Parse the table from Table resource.
     *
     * @param {string} tableName
     *   A fully-qualified path representing Table resource.
     * @returns {string} A string representing the table.
     */
    matchTableFromTableName(tableName) {
        return this.pathTemplates.tablePathTemplate.match(tableName).table;
    }
    /**
     * Terminate the gRPC channel and close the client.
     *
     * The client will no longer be usable and all future behavior is undefined.
     * @returns {Promise} A promise that resolves when the client is closed.
     */
    close() {
        if (this.bigtableInstanceAdminStub && !this._terminated) {
            return this.bigtableInstanceAdminStub.then(stub => {
                this._log.info('ending gRPC channel');
                this._terminated = true;
                stub.close();
                this.operationsClient.close();
            });
        }
        return Promise.resolve();
    }
}
exports.BigtableInstanceAdminClient = BigtableInstanceAdminClient;
