// @google-cloud/bigtable/build/src/v2/bigtable_client.js
"use strict";
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.BigtableClient = void 0;
const stream_1 = require("stream");
const jsonProtos = require("../../protos/protos.json");
const google_gax_1 = require("google-gax");
/**
 * Client JSON configuration object, loaded from
 * `src/v2/bigtable_client_config.json`.
 * This file defines retry strategy and timeouts for all API methods in this library.
 */
const gapicConfig = require("./bigtable_client_config.json");
const version = require('../../../package.json').version;
/**
 *  Service for reading from and writing to existing Bigtable tables.
 * @class
 * @memberof v2
 */
class BigtableClient {
    _terminated = false;
    _opts;
    _providedCustomServicePath;
    _gaxModule;
    _gaxGrpc;
    _protos;
    _defaults;
    _universeDomain;
    _servicePath;
    _log = google_gax_1.loggingUtils.log('bigtable');
    auth;
    descriptors = {
        page: {},
        stream: {},
        longrunning: {},
        batching: {},
    };
    warn;
    innerApiCalls;
    pathTemplates;
    bigtableStub;
    /**
     * Construct an instance of BigtableClient.
     *
     * @param {object} [options] - The configuration object.
     * The options accepted by the constructor are described in detail
     * in [this document](https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#creating-the-client-instance).
     * The common options are:
     * @param {object} [options.credentials] - Credentials object.
     * @param {string} [options.credentials.client_email]
     * @param {string} [options.credentials.private_key]
     * @param {string} [options.email] - Account email address. Required when
     *     using a .pem or .p12 keyFilename.
     * @param {string} [options.keyFilename] - Full path to the a .json, .pem, or
     *     .p12 key downloaded from the Google Developers Console. If you provide
     *     a path to a JSON file, the projectId option below is not necessary.
     *     NOTE: .pem and .p12 require you to specify options.email as well.
     * @param {number} [options.port] - The port on which to connect to
     *     the remote host.
     * @param {string} [options.projectId] - The project ID from the Google
     *     Developer's Console, e.g. 'grape-spaceship-123'. We will also check
     *     the environment variable GCLOUD_PROJECT for your project ID. If your
     *     app is running in an environment which supports
     *     {@link https://cloud.google.com/docs/authentication/application-default-credentials Application Default Credentials},
     *     your project ID will be detected automatically.
     * @param {string} [options.apiEndpoint] - The domain name of the
     *     API remote host.
     * @param {gax.ClientConfig} [options.clientConfig] - Client configuration override.
     *     Follows the structure of {@link gapicConfig}.
     * @param {boolean} [options.fallback] - Use HTTP/1.1 REST mode.
     *     For more information, please check the
     *     {@link https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#http11-rest-api-mode documentation}.
     * @param {gax} [gaxInstance]: loaded instance of `google-gax`. Useful if you
     *     need to avoid loading the default gRPC version and want to use the fallback
     *     HTTP implementation. Load only fallback version and pass it to the constructor:
     *     ```
     *     const gax = require('google-gax/build/src/fallback'); // avoids loading google-gax with gRPC
     *     const client = new BigtableClient({fallback: true}, gax);
     *     ```
     */
    constructor(opts, gaxInstance) {
        // Ensure that options include all the required fields.
        const staticMembers = this.constructor;
        if (opts?.universe_domain &&
            opts?.universeDomain &&
            opts?.universe_domain !== opts?.universeDomain) {
            throw new Error('Please set either universe_domain or universeDomain, but not both.');
        }
        const universeDomainEnvVar = typeof process === 'object' && typeof process.env === 'object'
            ? process.env['GOOGLE_CLOUD_UNIVERSE_DOMAIN']
            : undefined;
        this._universeDomain =
            opts?.universeDomain ??
                opts?.universe_domain ??
                universeDomainEnvVar ??
                'googleapis.com';
        this._servicePath = 'bigtable.' + this._universeDomain;
        const servicePath = opts?.servicePath || opts?.apiEndpoint || this._servicePath;
        this._providedCustomServicePath = !!(opts?.servicePath || opts?.apiEndpoint);
        const port = opts?.port || staticMembers.port;
        const clientConfig = opts?.clientConfig ?? {};
        const fallback = opts?.fallback ??
            (typeof window !== 'undefined' && typeof window?.fetch === 'function');
        opts = Object.assign({ servicePath, port, clientConfig, fallback }, opts);
        // Request numeric enum values if REST transport is used.
        opts.numericEnums = true;
        // If scopes are unset in options and we're connecting to a non-default endpoint, set scopes just in case.
        if (servicePath !== this._servicePath && !('scopes' in opts)) {
            opts['scopes'] = staticMembers.scopes;
        }
        // Load google-gax module synchronously if needed
        if (!gaxInstance) {
            gaxInstance = require('google-gax');
        }
        // Choose either gRPC or proto-over-HTTP implementation of google-gax.
        this._gaxModule = opts.fallback ? gaxInstance.fallback : gaxInstance;
        // Create a `gaxGrpc` object, with any grpc-specific options sent to the client.
        this._gaxGrpc = new this._gaxModule.GrpcClient(opts);
        // Save options to use in initialize() method.
        this._opts = opts;
        // Save the auth object to the client, for use by other methods.
        this.auth = this._gaxGrpc.auth;
        // Set useJWTAccessWithScope on the auth object.
        this.auth.useJWTAccessWithScope = true;
        // Set defaultServicePath on the auth object.
        this.auth.defaultServicePath = this._servicePath;
        // Set the default scopes in auth client if needed.
        if (servicePath === this._servicePath) {
            this.auth.defaultScopes = staticMembers.scopes;
        }
        // Determine the client header string.
        const clientHeader = [`gax/${this._gaxModule.version}`, `gapic/${version}`];
        if (typeof process === 'object' && 'versions' in process) {
            clientHeader.push(`gl-node/${process.versions.node}`);
        }
        else {
            clientHeader.push(`gl-web/${this._gaxModule.version}`);
        }
        if (!opts.fallback) {
            clientHeader.push(`grpc/${this._gaxGrpc.grpcVersion}`);
        }
        else {
            clientHeader.push(`rest/${this._gaxGrpc.grpcVersion}`);
        }
        if (opts.libName && opts.libVersion) {
            clientHeader.push(`${opts.libName}/${opts.libVersion}`);
        }
        // Load the applicable protos.
        this._protos = this._gaxGrpc.loadProtoJSON(jsonProtos);
        // This API contains "path templates"; forward-slash-separated
        // identifiers to uniquely identify resources within the API.
        // Create useful helper objects for these.
        this.pathTemplates = {
            authorizedViewPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/instances/{instance}/tables/{table}/authorizedViews/{authorized_view}'),
            instancePathTemplate: new this._gaxModule.PathTemplate('projects/{project}/instances/{instance}'),
            materializedViewPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/instances/{instance}/materializedViews/{materialized_view}'),
            tablePathTemplate: new this._gaxModule.PathTemplate('projects/{project}/instances/{instance}/tables/{table}'),
        };
        // Some of the methods on this service provide streaming responses.
        // Provide descriptors for these.
        this.descriptors.stream = {
            readRows: new this._gaxModule.StreamDescriptor(this._gaxModule.StreamType.SERVER_STREAMING, !!opts.fallback, !!opts.gaxServerStreamingRetries),
            sampleRowKeys: new this._gaxModule.StreamDescriptor(this._gaxModule.StreamType.SERVER_STREAMING, !!opts.fallback, !!opts.gaxServerStreamingRetries),
            mutateRows: new this._gaxModule.StreamDescriptor(this._gaxModule.StreamType.SERVER_STREAMING, !!opts.fallback, !!opts.gaxServerStreamingRetries),
            generateInitialChangeStreamPartitions: new this._gaxModule.StreamDescriptor(this._gaxModule.StreamType.SERVER_STREAMING, !!opts.fallback, !!opts.gaxServerStreamingRetries),
            readChangeStream: new this._gaxModule.StreamDescriptor(this._gaxModule.StreamType.SERVER_STREAMING, !!opts.fallback, !!opts.gaxServerStreamingRetries),
            executeQuery: new this._gaxModule.StreamDescriptor(this._gaxModule.StreamType.SERVER_STREAMING, !!opts.fallback, !!opts.gaxServerStreamingRetries),
        };
        // Put together the default options sent with requests.
        this._defaults = this._gaxGrpc.constructSettings('google.bigtable.v2.Bigtable', gapicConfig, opts.clientConfig || {}, { 'x-goog-api-client': clientHeader.join(' ') });
        // Set up a dictionary of "inner API calls"; the core implementation
        // of calling the API is handled in `google-gax`, with this code
        // merely providing the destination and request information.
        this.innerApiCalls = {};
        // Add a warn function to the client constructor so it can be easily tested.
        this.warn = this._gaxModule.warn;
    }
    /**
     * Initialize the client.
     * Performs asynchronous operations (such as authentication) and prepares the client.
     * This function will be called automatically when any class method is called for the
     * first time, but if you need to initialize it before calling an actual method,
     * feel free to call initialize() directly.
     *
     * You can await on this method if you want to make sure the client is initialized.
     *
     * @returns {Promise} A promise that resolves to an authenticated service stub.
     */
    initialize() {
        // If the client stub promise is already initialized, return immediately.
        if (this.bigtableStub) {
            return this.bigtableStub;
        }
        // Put together the "service stub" for
        // google.bigtable.v2.Bigtable.
        this.bigtableStub = this._gaxGrpc.createStub(this._opts.fallback
            ? this._protos.lookupService('google.bigtable.v2.Bigtable')
            : // eslint-disable-next-line @typescript-eslint/no-explicit-any
                this._protos.google.bigtable.v2.Bigtable, this._opts, this._providedCustomServicePath);
        // Iterate over each of the methods that the service provides
        // and create an API call method for each.
        const bigtableStubMethods = [
            'readRows',
            'sampleRowKeys',
            'mutateRow',
            'mutateRows',
            'checkAndMutateRow',
            'pingAndWarm',
            'readModifyWriteRow',
            'generateInitialChangeStreamPartitions',
            'readChangeStream',
            'prepareQuery',
            'executeQuery',
        ];
        for (const methodName of bigtableStubMethods) {
            const callPromise = this.bigtableStub.then(stub => (...args) => {
                if (this._terminated) {
                    if (methodName in this.descriptors.stream) {
                        const stream = new stream_1.PassThrough({ objectMode: true });
                        setImmediate(() => {
                            stream.emit('error', new this._gaxModule.GoogleError('The client has already been closed.'));
                        });
                        return stream;
                    }
                    return Promise.reject('The client has already been closed.');
                }
                const func = stub[methodName];
                return func.apply(stub, args);
            }, (err) => () => {
                throw err;
            });
            const descriptor = this.descriptors.stream[methodName] || undefined;
            const apiCall = this._gaxModule.createApiCall(callPromise, this._defaults[methodName], descriptor, this._opts.fallback);
            this.innerApiCalls[methodName] = apiCall;
        }
        return this.bigtableStub;
    }
    /**
     * The DNS address for this API service.
     * @deprecated Use the apiEndpoint method of the client instance.
     * @returns {string} The DNS address for this service.
     */
    static get servicePath() {
        if (typeof process === 'object' &&
            typeof process.emitWarning === 'function') {
            process.emitWarning('Static servicePath is deprecated, please use the instance method instead.', 'DeprecationWarning');
        }
        return 'bigtable.googleapis.com';
    }
    /**
     * The DNS address for this API service - same as servicePath.
     * @deprecated Use the apiEndpoint method of the client instance.
     * @returns {string} The DNS address for this service.
     */
    static get apiEndpoint() {
        if (typeof process === 'object' &&
            typeof process.emitWarning === 'function') {
            process.emitWarning('Static apiEndpoint is deprecated, please use the instance method instead.', 'DeprecationWarning');
        }
        return 'bigtable.googleapis.com';
    }
    /**
     * The DNS address for this API service.
     * @returns {string} The DNS address for this service.
     */
    get apiEndpoint() {
        return this._servicePath;
    }
    get universeDomain() {
        return this._universeDomain;
    }
    /**
     * The port for this API service.
     * @returns {number} The default port for this service.
     */
    static get port() {
        return 443;
    }
    /**
     * The scopes needed to make gRPC calls for every method defined
     * in this service.
     * @returns {string[]} List of default scopes.
     */
    static get scopes() {
        return [
            'https://www.googleapis.com/auth/bigtable.data',
            'https://www.googleapis.com/auth/bigtable.data.readonly',
            'https://www.googleapis.com/auth/cloud-bigtable.data',
            'https://www.googleapis.com/auth/cloud-bigtable.data.readonly',
            'https://www.googleapis.com/auth/cloud-platform',
            'https://www.googleapis.com/auth/cloud-platform.read-only',
        ];
    }
    /**
     * Return the project ID used by this class.
     * @returns {Promise} A promise that resolves to string containing the project ID.
     */
    getProjectId(callback) {
        if (callback) {
            this.auth.getProjectId(callback);
            return;
        }
        return this.auth.getProjectId();
    }
    mutateRow(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const routingParameter = {};
        {
            const fieldValue = request.tableName;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<table_name>projects/[^/]+/instances/[^/]+/tables/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['table_name'] ?? fieldValue;
                    Object.assign(routingParameter, { table_name: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.appProfileId;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<app_profile_id>.*)'));
                if (match) {
                    const parameterValue = match.groups?.['app_profile_id'] ?? fieldValue;
                    Object.assign(routingParameter, { app_profile_id: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.authorizedViewName;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<authorized_view_name>projects/[^/]+/instances/[^/]+/tables/[^/]+/authorizedViews/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['authorized_view_name'] ?? fieldValue;
                    Object.assign(routingParameter, {
                        authorized_view_name: parameterValue,
                    });
                }
            }
        }
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams(routingParameter);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('mutateRow request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('mutateRow response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .mutateRow(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('mutateRow response %j', response);
            return [response, options, rawResponse];
        });
    }
    checkAndMutateRow(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const routingParameter = {};
        {
            const fieldValue = request.tableName;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<table_name>projects/[^/]+/instances/[^/]+/tables/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['table_name'] ?? fieldValue;
                    Object.assign(routingParameter, { table_name: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.appProfileId;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<app_profile_id>.*)'));
                if (match) {
                    const parameterValue = match.groups?.['app_profile_id'] ?? fieldValue;
                    Object.assign(routingParameter, { app_profile_id: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.authorizedViewName;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<authorized_view_name>projects/[^/]+/instances/[^/]+/tables/[^/]+/authorizedViews/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['authorized_view_name'] ?? fieldValue;
                    Object.assign(routingParameter, {
                        authorized_view_name: parameterValue,
                    });
                }
            }
        }
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams(routingParameter);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('checkAndMutateRow request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('checkAndMutateRow response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .checkAndMutateRow(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('checkAndMutateRow response %j', response);
            return [response, options, rawResponse];
        });
    }
    pingAndWarm(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const routingParameter = {};
        {
            const fieldValue = request.name;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<name>projects/[^/]+/instances/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['name'] ?? fieldValue;
                    Object.assign(routingParameter, { name: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.appProfileId;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<app_profile_id>.*)'));
                if (match) {
                    const parameterValue = match.groups?.['app_profile_id'] ?? fieldValue;
                    Object.assign(routingParameter, { app_profile_id: parameterValue });
                }
            }
        }
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams(routingParameter);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('pingAndWarm request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('pingAndWarm response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .pingAndWarm(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('pingAndWarm response %j', response);
            return [response, options, rawResponse];
        });
    }
    readModifyWriteRow(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const routingParameter = {};
        {
            const fieldValue = request.tableName;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<table_name>projects/[^/]+/instances/[^/]+/tables/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['table_name'] ?? fieldValue;
                    Object.assign(routingParameter, { table_name: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.appProfileId;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<app_profile_id>.*)'));
                if (match) {
                    const parameterValue = match.groups?.['app_profile_id'] ?? fieldValue;
                    Object.assign(routingParameter, { app_profile_id: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.authorizedViewName;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<authorized_view_name>projects/[^/]+/instances/[^/]+/tables/[^/]+/authorizedViews/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['authorized_view_name'] ?? fieldValue;
                    Object.assign(routingParameter, {
                        authorized_view_name: parameterValue,
                    });
                }
            }
        }
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams(routingParameter);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('readModifyWriteRow request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('readModifyWriteRow response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .readModifyWriteRow(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('readModifyWriteRow response %j', response);
            return [response, options, rawResponse];
        });
    }
    prepareQuery(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const routingParameter = {};
        {
            const fieldValue = request.instanceName;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<name>projects/[^/]+/instances/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['name'] ?? fieldValue;
                    Object.assign(routingParameter, { name: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.appProfileId;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<app_profile_id>.*)'));
                if (match) {
                    const parameterValue = match.groups?.['app_profile_id'] ?? fieldValue;
                    Object.assign(routingParameter, { app_profile_id: parameterValue });
                }
            }
        }
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams(routingParameter);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('prepareQuery request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('prepareQuery response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls
            .prepareQuery(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('prepareQuery response %j', response);
            return [response, options, rawResponse];
        });
    }
    /**
     * Streams back the contents of all requested rows in key order, optionally
     * applying the same Reader filter to each. Depending on their size,
     * rows and cells may be broken up across multiple responses, but
     * atomicity of each row will still be preserved. See the
     * ReadRowsResponse documentation for details.
     *
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} [request.tableName]
     *   Optional. The unique name of the table from which to read.
     *
     *   Values are of the form
     *   `projects/<project>/instances/<instance>/tables/<table>`.
     * @param {string} [request.authorizedViewName]
     *   Optional. The unique name of the AuthorizedView from which to read.
     *
     *   Values are of the form
     *   `projects/<project>/instances/<instance>/tables/<table>/authorizedViews/<authorized_view>`.
     * @param {string} [request.materializedViewName]
     *   Optional. The unique name of the MaterializedView from which to read.
     *
     *   Values are of the form
     *   `projects/<project>/instances/<instance>/materializedViews/<materialized_view>`.
     * @param {string} request.appProfileId
     *   This value specifies routing for replication. If not specified, the
     *   "default" application profile will be used.
     * @param {google.bigtable.v2.RowSet} request.rows
     *   The row keys and/or ranges to read sequentially. If not specified, reads
     *   from all rows.
     * @param {google.bigtable.v2.RowFilter} request.filter
     *   The filter to apply to the contents of the specified row(s). If unset,
     *   reads the entirety of each row.
     * @param {number} request.rowsLimit
     *   The read will stop after committing to N rows' worth of results. The
     *   default (zero) is to return all results.
     * @param {google.bigtable.v2.ReadRowsRequest.RequestStatsView} request.requestStatsView
     *   The view into RequestStats, as described above.
     * @param {boolean} request.reversed
     *   Experimental API - Please note that this API is currently experimental
     *   and can change in the future.
     *
     *   Return rows in lexiographical descending order of the row keys. The row
     *   contents will not be affected by this flag.
     *
     *   Example result set:
     *
     *       [
     *         {key: "k2", "f:col1": "v1", "f:col2": "v1"},
     *         {key: "k1", "f:col1": "v2", "f:col2": "v2"}
     *       ]
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits {@link protos.google.bigtable.v2.ReadRowsResponse|ReadRowsResponse} on 'data' event.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#server-streaming | documentation }
     *   for more details and examples.
     */
    readRows(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const routingParameter = {};
        {
            const fieldValue = request.tableName;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<table_name>projects/[^/]+/instances/[^/]+/tables/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['table_name'] ?? fieldValue;
                    Object.assign(routingParameter, { table_name: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.appProfileId;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<app_profile_id>.*)'));
                if (match) {
                    const parameterValue = match.groups?.['app_profile_id'] ?? fieldValue;
                    Object.assign(routingParameter, { app_profile_id: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.authorizedViewName;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<authorized_view_name>projects/[^/]+/instances/[^/]+/tables/[^/]+/authorizedViews/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['authorized_view_name'] ?? fieldValue;
                    Object.assign(routingParameter, {
                        authorized_view_name: parameterValue,
                    });
                }
            }
        }
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams(routingParameter);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('readRows stream %j', options);
        return this.innerApiCalls.readRows(request, options);
    }
    /**
     * Returns a sample of row keys in the table. The returned row keys will
     * delimit contiguous sections of the table of approximately equal size,
     * which can be used to break up the data for distributed tasks like
     * mapreduces.
     *
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} [request.tableName]
     *   Optional. The unique name of the table from which to sample row keys.
     *
     *   Values are of the form
     *   `projects/<project>/instances/<instance>/tables/<table>`.
     * @param {string} [request.authorizedViewName]
     *   Optional. The unique name of the AuthorizedView from which to sample row
     *   keys.
     *
     *   Values are of the form
     *   `projects/<project>/instances/<instance>/tables/<table>/authorizedViews/<authorized_view>`.
     * @param {string} [request.materializedViewName]
     *   Optional. The unique name of the MaterializedView from which to read.
     *
     *   Values are of the form
     *   `projects/<project>/instances/<instance>/materializedViews/<materialized_view>`.
     * @param {string} request.appProfileId
     *   This value specifies routing for replication. If not specified, the
     *   "default" application profile will be used.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits {@link protos.google.bigtable.v2.SampleRowKeysResponse|SampleRowKeysResponse} on 'data' event.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#server-streaming | documentation }
     *   for more details and examples.
     */
    sampleRowKeys(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const routingParameter = {};
        {
            const fieldValue = request.tableName;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<table_name>projects/[^/]+/instances/[^/]+/tables/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['table_name'] ?? fieldValue;
                    Object.assign(routingParameter, { table_name: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.appProfileId;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<app_profile_id>.*)'));
                if (match) {
                    const parameterValue = match.groups?.['app_profile_id'] ?? fieldValue;
                    Object.assign(routingParameter, { app_profile_id: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.authorizedViewName;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<authorized_view_name>projects/[^/]+/instances/[^/]+/tables/[^/]+/authorizedViews/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['authorized_view_name'] ?? fieldValue;
                    Object.assign(routingParameter, {
                        authorized_view_name: parameterValue,
                    });
                }
            }
        }
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams(routingParameter);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('sampleRowKeys stream %j', options);
        return this.innerApiCalls.sampleRowKeys(request, options);
    }
    /**
     * Mutates multiple rows in a batch. Each individual row is mutated
     * atomically as in MutateRow, but the entire batch is not executed
     * atomically.
     *
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} [request.tableName]
     *   Optional. The unique name of the table to which the mutations should be
     *   applied.
     *
     *   Values are of the form
     *   `projects/<project>/instances/<instance>/tables/<table>`.
     * @param {string} [request.authorizedViewName]
     *   Optional. The unique name of the AuthorizedView to which the mutations
     *   should be applied.
     *
     *   Values are of the form
     *   `projects/<project>/instances/<instance>/tables/<table>/authorizedViews/<authorized_view>`.
     * @param {string} request.appProfileId
     *   This value specifies routing for replication. If not specified, the
     *   "default" application profile will be used.
     * @param {number[]} request.entries
     *   Required. The row keys and corresponding mutations to be applied in bulk.
     *   Each entry is applied as an atomic mutation, but the entries may be
     *   applied in arbitrary order (even between entries for the same row).
     *   At least one entry must be specified, and in total the entries can
     *   contain at most 100000 mutations.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits {@link protos.google.bigtable.v2.MutateRowsResponse|MutateRowsResponse} on 'data' event.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#server-streaming | documentation }
     *   for more details and examples.
     */
    mutateRows(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const routingParameter = {};
        {
            const fieldValue = request.tableName;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<table_name>projects/[^/]+/instances/[^/]+/tables/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['table_name'] ?? fieldValue;
                    Object.assign(routingParameter, { table_name: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.appProfileId;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<app_profile_id>.*)'));
                if (match) {
                    const parameterValue = match.groups?.['app_profile_id'] ?? fieldValue;
                    Object.assign(routingParameter, { app_profile_id: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.authorizedViewName;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<authorized_view_name>projects/[^/]+/instances/[^/]+/tables/[^/]+/authorizedViews/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['authorized_view_name'] ?? fieldValue;
                    Object.assign(routingParameter, {
                        authorized_view_name: parameterValue,
                    });
                }
            }
        }
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams(routingParameter);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('mutateRows stream %j', options);
        return this.innerApiCalls.mutateRows(request, options);
    }
    /**
     * NOTE: This API is intended to be used by Apache Beam BigtableIO.
     * Returns the current list of partitions that make up the table's
     * change stream. The union of partitions will cover the entire keyspace.
     * Partitions can be read with `ReadChangeStream`.
     *
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.tableName
     *   Required. The unique name of the table from which to get change stream
     *   partitions. Values are of the form
     *   `projects/<project>/instances/<instance>/tables/<table>`.
     *   Change streaming must be enabled on the table.
     * @param {string} request.appProfileId
     *   This value specifies routing for replication. If not specified, the
     *   "default" application profile will be used.
     *   Single cluster routing must be configured on the profile.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits {@link protos.google.bigtable.v2.GenerateInitialChangeStreamPartitionsResponse|GenerateInitialChangeStreamPartitionsResponse} on 'data' event.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#server-streaming | documentation }
     *   for more details and examples.
     */
    generateInitialChangeStreamPartitions(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                table_name: request.tableName ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('generateInitialChangeStreamPartitions stream %j', options);
        return this.innerApiCalls.generateInitialChangeStreamPartitions(request, options);
    }
    /**
     * NOTE: This API is intended to be used by Apache Beam BigtableIO.
     * Reads changes from a table's change stream. Changes will
     * reflect both user-initiated mutations and mutations that are caused by
     * garbage collection.
     *
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.tableName
     *   Required. The unique name of the table from which to read a change stream.
     *   Values are of the form
     *   `projects/<project>/instances/<instance>/tables/<table>`.
     *   Change streaming must be enabled on the table.
     * @param {string} request.appProfileId
     *   This value specifies routing for replication. If not specified, the
     *   "default" application profile will be used.
     *   Single cluster routing must be configured on the profile.
     * @param {google.bigtable.v2.StreamPartition} request.partition
     *   The partition to read changes from.
     * @param {google.protobuf.Timestamp} request.startTime
     *   Start reading the stream at the specified timestamp. This timestamp must
     *   be within the change stream retention period, less than or equal to the
     *   current time, and after change stream creation, whichever is greater.
     *   This value is inclusive and will be truncated to microsecond granularity.
     * @param {google.bigtable.v2.StreamContinuationTokens} request.continuationTokens
     *   Tokens that describe how to resume reading a stream where reading
     *   previously left off. If specified, changes will be read starting at the
     *   the position. Tokens are delivered on the stream as part of `Heartbeat`
     *   and `CloseStream` messages.
     *
     *   If a single token is provided, the token’s partition must exactly match
     *   the request’s partition. If multiple tokens are provided, as in the case
     *   of a partition merge, the union of the token partitions must exactly
     *   cover the request’s partition. Otherwise, INVALID_ARGUMENT will be
     *   returned.
     * @param {google.protobuf.Timestamp} request.endTime
     *   If specified, OK will be returned when the stream advances beyond
     *   this time. Otherwise, changes will be continuously delivered on the stream.
     *   This value is inclusive and will be truncated to microsecond granularity.
     * @param {google.protobuf.Duration} request.heartbeatDuration
     *   If specified, the duration between `Heartbeat` messages on the stream.
     *   Otherwise, defaults to 5 seconds.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits {@link protos.google.bigtable.v2.ReadChangeStreamResponse|ReadChangeStreamResponse} on 'data' event.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#server-streaming | documentation }
     *   for more details and examples.
     */
    readChangeStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                table_name: request.tableName ?? '',
            });
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('readChangeStream stream %j', options);
        return this.innerApiCalls.readChangeStream(request, options);
    }
    /**
     * Executes a SQL query against a particular Bigtable instance.
     *
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.instanceName
     *   Required. The unique name of the instance against which the query should be
     *   executed.
     *   Values are of the form `projects/<project>/instances/<instance>`
     * @param {string} [request.appProfileId]
     *   Optional. This value specifies routing for replication. If not specified,
     *   the `default` application profile will be used.
     * @param {string} request.query
     *   Required. The query string.
     *
     *   Exactly one of `query` and `prepared_query` is required. Setting both
     *   or neither is an `INVALID_ARGUMENT`.
     * @param {Buffer} request.preparedQuery
     *   A prepared query that was returned from `PrepareQueryResponse`.
     *
     *   Exactly one of `query` and `prepared_query` is required. Setting both
     *   or neither is an `INVALID_ARGUMENT`.
     *
     *   Setting this field also places restrictions on several other fields:
     *   - `data_format` must be empty.
     *   - `validate_only` must be false.
     *   - `params` must match the `param_types` set in the `PrepareQueryRequest`.
     * @param {google.bigtable.v2.ProtoFormat} request.protoFormat
     *   Protocol buffer format as described by ProtoSchema and ProtoRows
     *   messages.
     * @param {Buffer} [request.resumeToken]
     *   Optional. If this request is resuming a previously interrupted query
     *   execution, `resume_token` should be copied from the last
     *   PartialResultSet yielded before the interruption. Doing this
     *   enables the query execution to resume where the last one left
     *   off.
     *   The rest of the request parameters must exactly match the
     *   request that yielded this token. Otherwise the request will fail.
     * @param {number[]} request.params
     *   Required. params contains string type keys and Bigtable type values that
     *   bind to placeholders in the query string. In query string, a parameter
     *   placeholder consists of the
     *   `@` character followed by the parameter name (for example, `@firstName`) in
     *   the query string.
     *
     *   For example, if
     *   `params["firstName"] = bytes_value: "foo" type {bytes_type {}}`
     *   then `@firstName` will be replaced with googlesql bytes value "foo" in the
     *   query string during query evaluation.
     *
     *   If `Value.kind` is not set, the value is treated as a NULL value of the
     *   given type. For example, if
     *   `params["firstName"] = type {string_type {}}`
     *   then `@firstName` will be replaced with googlesql null string.
     *
     *   If `query` is set, any empty `Value.type` in the map will be rejected with
     *   `INVALID_ARGUMENT`.
     *
     *   If `prepared_query` is set, any empty `Value.type` in the map will be
     *   inferred from the `param_types` in the `PrepareQueryRequest`. Any non-empty
     *   `Value.type` must match the corresponding `param_types` entry, or be
     *   rejected with `INVALID_ARGUMENT`.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits {@link protos.google.bigtable.v2.ExecuteQueryResponse|ExecuteQueryResponse} on 'data' event.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#server-streaming | documentation }
     *   for more details and examples.
     */
    executeQuery(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const routingParameter = {};
        {
            const fieldValue = request.instanceName;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<name>projects/[^/]+/instances/[^/]+)'));
                if (match) {
                    const parameterValue = match.groups?.['name'] ?? fieldValue;
                    Object.assign(routingParameter, { name: parameterValue });
                }
            }
        }
        {
            const fieldValue = request.appProfileId;
            if (fieldValue !== undefined && fieldValue !== null) {
                const match = fieldValue
                    .toString()
                    .match(RegExp('(?<app_profile_id>.*)'));
                if (match) {
                    const parameterValue = match.groups?.['app_profile_id'] ?? fieldValue;
                    Object.assign(routingParameter, { app_profile_id: parameterValue });
                }
            }
        }
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams(routingParameter);
        this.initialize().catch(err => {
            throw err;
        });
        this._log.info('executeQuery stream %j', options);
        return this.innerApiCalls.executeQuery(request, options);
    }
    // --------------------
    // -- Path templates --
    // --------------------
    /**
     * Return a fully-qualified authorizedView resource name string.
     *
     * @param {string} project
     * @param {string} instance
     * @param {string} table
     * @param {string} authorized_view
     * @returns {string} Resource name string.
     */
    authorizedViewPath(project, instance, table, authorizedView) {
        return this.pathTemplates.authorizedViewPathTemplate.render({
            project: project,
            instance: instance,
            table: table,
            authorized_view: authorizedView,
        });
    }
    /**
     * Parse the project from AuthorizedView resource.
     *
     * @param {string} authorizedViewName
     *   A fully-qualified path representing AuthorizedView resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromAuthorizedViewName(authorizedViewName) {
        return this.pathTemplates.authorizedViewPathTemplate.match(authorizedViewName).project;
    }
    /**
     * Parse the instance from AuthorizedView resource.
     *
     * @param {string} authorizedViewName
     *   A fully-qualified path representing AuthorizedView resource.
     * @returns {string} A string representing the instance.
     */
    matchInstanceFromAuthorizedViewName(authorizedViewName) {
        return this.pathTemplates.authorizedViewPathTemplate.match(authorizedViewName).instance;
    }
    /**
     * Parse the table from AuthorizedView resource.
     *
     * @param {string} authorizedViewName
     *   A fully-qualified path representing AuthorizedView resource.
     * @returns {string} A string representing the table.
     */
    matchTableFromAuthorizedViewName(authorizedViewName) {
        return this.pathTemplates.authorizedViewPathTemplate.match(authorizedViewName).table;
    }
    /**
     * Parse the authorized_view from AuthorizedView resource.
     *
     * @param {string} authorizedViewName
     *   A fully-qualified path representing AuthorizedView resource.
     * @returns {string} A string representing the authorized_view.
     */
    matchAuthorizedViewFromAuthorizedViewName(authorizedViewName) {
        return this.pathTemplates.authorizedViewPathTemplate.match(authorizedViewName).authorized_view;
    }
    /**
     * Return a fully-qualified instance resource name string.
     *
     * @param {string} project
     * @param {string} instance
     * @returns {string} Resource name string.
     */
    instancePath(project, instance) {
        return this.pathTemplates.instancePathTemplate.render({
            project: project,
            instance: instance,
        });
    }
    /**
     * Parse the project from Instance resource.
     *
     * @param {string} instanceName
     *   A fully-qualified path representing Instance resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromInstanceName(instanceName) {
        return this.pathTemplates.instancePathTemplate.match(instanceName).project;
    }
    /**
     * Parse the instance from Instance resource.
     *
     * @param {string} instanceName
     *   A fully-qualified path representing Instance resource.
     * @returns {string} A string representing the instance.
     */
    matchInstanceFromInstanceName(instanceName) {
        return this.pathTemplates.instancePathTemplate.match(instanceName).instance;
    }
    /**
     * Return a fully-qualified materializedView resource name string.
     *
     * @param {string} project
     * @param {string} instance
     * @param {string} materialized_view
     * @returns {string} Resource name string.
     */
    materializedViewPath(project, instance, materializedView) {
        return this.pathTemplates.materializedViewPathTemplate.render({
            project: project,
            instance: instance,
            materialized_view: materializedView,
        });
    }
    /**
     * Parse the project from MaterializedView resource.
     *
     * @param {string} materializedViewName
     *   A fully-qualified path representing MaterializedView resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromMaterializedViewName(materializedViewName) {
        return this.pathTemplates.materializedViewPathTemplate.match(materializedViewName).project;
    }
    /**
     * Parse the instance from MaterializedView resource.
     *
     * @param {string} materializedViewName
     *   A fully-qualified path representing MaterializedView resource.
     * @returns {string} A string representing the instance.
     */
    matchInstanceFromMaterializedViewName(materializedViewName) {
        return this.pathTemplates.materializedViewPathTemplate.match(materializedViewName).instance;
    }
    /**
     * Parse the materialized_view from MaterializedView resource.
     *
     * @param {string} materializedViewName
     *   A fully-qualified path representing MaterializedView resource.
     * @returns {string} A string representing the materialized_view.
     */
    matchMaterializedViewFromMaterializedViewName(materializedViewName) {
        return this.pathTemplates.materializedViewPathTemplate.match(materializedViewName).materialized_view;
    }
    /**
     * Return a fully-qualified table resource name string.
     *
     * @param {string} project
     * @param {string} instance
     * @param {string} table
     * @returns {string} Resource name string.
     */
    tablePath(project, instance, table) {
        return this.pathTemplates.tablePathTemplate.render({
            project: project,
            instance: instance,
            table: table,
        });
    }
    /**
     * Parse the project from Table resource.
     *
     * @param {string} tableName
     *   A fully-qualified path representing Table resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromTableName(tableName) {
        return this.pathTemplates.tablePathTemplate.match(tableName).project;
    }
    /**
     * Parse the instance from Table resource.
     *
     * @param {string} tableName
     *   A fully-qualified path representing Table resource.
     * @returns {string} A string representing the instance.
     */
    matchInstanceFromTableName(tableName) {
        return this.pathTemplates.tablePathTemplate.match(tableName).instance;
    }
    /**
     * Parse the table from Table resource.
     *
     * @param {string} tableName
     *   A fully-qualified path representing Table resource.
     * @returns {string} A string representing the table.
     */
    matchTableFromTableName(tableName) {
        return this.pathTemplates.tablePathTemplate.match(tableName).table;
    }
    /**
     * Terminate the gRPC channel and close the client.
     *
     * The client will no longer be usable and all future behavior is undefined.
     * @returns {Promise} A promise that resolves when the client is closed.
     */
    close() {
        if (this.bigtableStub && !this._terminated) {
            return this.bigtableStub.then(stub => {
                this._log.info('ending gRPC channel');
                this._terminated = true;
                stub.close();
            });
        }
        return Promise.resolve();
    }
}
exports.BigtableClient = BigtableClient;
